Sharing: add a Mastodon sharing button

jeherve opened the pull request:
Fixes #7148

## Proposed changes:

This introduces a new Mastodon sharing button:

<img width="470" alt="image" src="https://user-images.githubusercontent.com/426388/216085445-333f505e-f8ae-4cb2-960f-a9420cc07792.png">
<img width="463" alt="Screenshot 2023-02-13 at 14 56 17" src="https://user-images.githubusercontent.com/426388/218478046-35a684ec-ac5b-452e-b074-a13aee2a1048.png">
<img width="459" alt="Screenshot 2023-02-13 at 14 58 56" src="https://user-images.githubusercontent.com/426388/218478076-3847a103-c1e5-4632-902f-6c00616a5464.png">

### Notes

**Class name**

In the past, we once had to give a unique name to a sharing service's class (`Jetpack_Share_WhatsApp`) to avoid conflicting with an existing plugin that already extended the sharing module. I looked around a bit, but could not find any plugin that does that for Mastodon today; folks who want a sharing button on their site today use the "custom sharing button" option we provide in the sharing settings.

**Mastodon support**

This will only support Mastodon. There are many other ActivityPub clients out there, but Mastodon is the biggest one so far, and is one that does offer a standardized "sharing" interface via the web.

In the future, if other clients become more and more popular, we can consider supporting those as well.

**Custom instance support**

We do, however, want to allow site visitors to select on which instance they want to post. The sharing button will consequently need to support inputing an instance name.

**Publicize**

This adds a sharing button, to allow site visitors to manually share blog posts to a Mastodon instance. For automated posting to Mastodon, we track requests and will post updates in #7790.

Until this is implemented, one can use [the ActivityPub plugin](https://wordpress.org/plugins/activitypub/) to have your site publish all your posts to the Fediverse. You can then Boost those posts from your Mastodon / Fediverse account if you have one.

### To do

- [x] Add a first modal view where one can enter their instance URL (or maybe rather their full Mastodon username, and we'd pull the instance name from that.
- [x] Consider a different default message, maybe pulling in a post's tags as hashtags.
- [x] Add a filter so site owners can customize that default message.
- [x] Ensure the Mastodon sharing button can also be added and managed from the Calypso UI (at `https://wordpress.com/marketing/sharing-buttons/yoursite.com`). This will require a Calypso PR: https://github.com/Automattic/wp-calypso/pull/72876

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

* pewaj1-3x-p2

## Does this pull request change what data or activity we track or use?
* No

## Testing instructions:

You'll want to start with a site that's connected to WordPress.com, or where you've enabled [Jetpack's Offline Mode](https://jetpack.com/support/development-mode/).

* Go to Jetpack > Settings > Sharing
* Enable the sharing buttons
* Go to Settings > Sharing
* Drag a Mastodon sharing button to one of the sharing areas.
    *  You will want with different button styles.
* Check that the button works on the frontend.

jeherve committed:
Social Logos: update with most recent changes

This updates the icon font to match the changes that were introduced in the most recent Social Logos updates. The last update specifically introduced a Mastodon logo.

Matching package update PR: #28685

jeherve committed:
Sharing: add initial version of a Mastodon sharing button.

This is a first version, that will need to be amended. The modal will first need to ask what instance you want to post to, so the sharing URL can be modified.

jeherve committed:
Avoid whitespace noise.

jeherve committed:
Add filter for the default message

jeherve committed:
Make the modal a tiny bit smaller

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into add/mastodon-sharing-button

jeherve committed:
Add first basic instance prompt

- Add new page template to display just the prompt.
- Display the page template when no POST data is provided.
- When we do have POST data, use that to start redirecting to the right Mastodon instance's page.

jeherve committed:
Ensure we use the right URL.

jeherve committed:
Fix display of the placeholer, and add title

jeherve committed:
Style form

jeherve committed:
Themes can impact the font family, let's not allow that.

jeherve committed:
Add the icon to the official button style too

jeherve committed:
Add post tags to the shared message

This introduces a new method we can use for other services in the future.

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into add/mastodon-sharing-button

jeherve commented:
@kraftbj As an active Mastodon user, I was wondering if you could give this a try, and let me know what you think of it?

It is a bit heavy with a full template page for the interstitial, but I figured it may be better that way.

jeherve committed:
Add sharing ID to filter to make it more useful

kraftbj commented:
I didn't dig into any of this, just initial thoughts:


![2023-02-08 at 11 27 AM](https://user-images.githubusercontent.com/88897/217606280-5cadcd0a-0f48-452e-ba12-ecb370d1d40e.png)

The icon didn't resolve.

![2023-02-08 at 11 27 AM](https://user-images.githubusercontent.com/88897/217606384-28451312-6d0a-4d46-a942-97577987a6e9.png)

I like this. It *looks* right. The downside is if Mastodon's default theme changes, we need to keep up. But, it feels good to me.

![2023-02-08 at 11 28 AM](https://user-images.githubusercontent.com/88897/217606617-6cd31e9a-9120-484e-81b2-24f6a4398480.png)

Cool, but the tags should camel case or hyphen or something( see `#distributed work`)

jeherve committed:
Camel case tags and remove spaces and apostrophes

See https://github.com/Automattic/jetpack/pull/28694#issuecomment-1422988427

jeherve commented:
> The icon didn't resolve.

This is interesting, I'm not sure why this would happen. Do you happen to have another plugin that would load its own version of the social logos, maybe?
Do you see any errors in the console? Does this happen in a specific browser?

Edit: I see it on your site now. I should have checked :)

```
kraft.blog/:1 Failed to decode downloaded font: https://kraft.blog/wp-content/plugins/jetpack-dev/_inc/social-logos/social-logos.eot?5d3b4da4f6c2460dd842dbf9e0992ea6
kraft.blog/:1 OTS parsing error: invalid sfntVersion: 1882128384
```

I'll look into that. This reminds me of #7148.

> Cool, but the tags should camel case or hyphen or something( see #distributed work)

Ah yes, that's a good catch. I fixed that in abad01a7a846a0400633e952abfbaef982bbbad0

jeherve commented:
> I'll look into that. This reminds me of #7148.

Do you have a way to disable Fontawesome in the Genesis Blocks plugin? In this linked issue, there was a suspicion that having multiple `@font-face` declarations could cause issues. I wonder if that could be it.

Although I do not really understand why this is only popping up now for you, and only for that one icon. If the browser "cannot decode downloaded font", why do we see the Twitter logo just fine? 🤔

jeherve committed:
Remove .eot handling from Social Logos

Fixes #7148

We really only need `woff` and `woff2` these days. We'll keep `ttf` as well for good measure. `eot` was really only useful for IE7 and IE8, we do not need it anymore.

Reference:
-  https://css-tricks.com/snippets/css/using-font-face-in-css/#aa-practical-level-of-browser-support

jeherve commented:
> I'll look into that. This reminds me of #7148.

I think I'll just take this opportunity to simplify things and get rid of eot. That's what we had considered back in 2016. We really only need `woff` and `woff2` these days, but I'll leave `ttf` as well for good measure. This dates back to times when we needed to support IE11. We do not need to worry about eot anymore. Removing it fixes the issue on your site, in Firefox, Chrome, Safari, and Edge.

I just pushed 84c2ecee17ef6d313c296959d0f21c06ead04950 to remove eot.

> Things are [shifting heavily toward WOFF](http://caniuse.com/#feat=woff) and [WOFF 2](http://caniuse.com/#feat=woff2)
-- https://css-tricks.com/snippets/css/using-font-face-in-css/#aa-practical-level-of-browser-support

Since that change is not really directly related to this PR, I'm happy to pull it into its own PR => #28849.

jeherve committed:
Add icon font update changelog

jeherve committed:
Update minified file too.

jeherve committed:
Update minified Social Logos CSS

jeherve committed:
Remove stray closing script tag

kraftbj committed:
Merge remote-tracking branch 'origin/trunk' into add/mastodon-sharing-button

jeherve committed:
Update changelogs now that #28849 is merged

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into add/mastodon-sharing-button

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into add/mastodon-sharing-button

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into add/mastodon-sharing-button

jeherve committed:
Style the form via inline CSS in the template directly.

This has a few advantages:

* We avoid loading extra CSS on pages that do not need it, and on sites that may not use the Mastodon sharing button.
* By not relying on `wp_head()` to bring in the styles to the template, we ensure that different themes do not impact the look of the modal.
* In the future, if we need to update the Mastodon styles, we can do it all in one place.

This has some inconvenients:

* We now have a hardcoded Mastodon icon in there, we should not forget to update it if the Mastodon icon ever changes.
* Since we have no `wp_head()` in that template anymore, it is now harder for folks to customize that page if they want to. I may add in hooks to make that easier (although I do not want if customizing that page is that important).

jeherve committed:
Add canonical link

jeherve committed:
Do not pass referrer information

This should add a layer of privacy to the shares: instance admins will not be able to filter by referrer to see links that were shared by the users on their instance.

Obviously, since the post URL is being shared it is included in the query string so there is no total privacy there, but there isn't much we can do about that.

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into add/mastodon-sharing-button

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into add/mastodon-sharing-button

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into add/mastodon-sharing-button

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into add/mastodon-sharing-button

anomiex commented on projects/plugins/jetpack/modules/sharedaddy/services/class-jetpack-mastodon-modal.php:
```diff
@@ -0,0 +1,195 @@
+	 *
+	 * @var string
+	 */
+	private static $svg = '<svg width="261" height="66" viewBox="0 0 261 66" fill="none" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
```
Would it make sense to put this image in the images directory, and read it with `file_get_contents()` if necessary instead of embedding it in the PHP file?

jeherve replied:
Right now I don't really imagine that logo will be used anywhere else, so I'm okay with hardcoding it in here. If we ever start using a full logo like that, I'll definitely move it out to its own file.

anomiex replied:
I was thinking more that it would be easier to deal with as an actual SVG file than as a long PHP string.

anomiex commented on projects/plugins/jetpack/modules/sharedaddy/services/class-jetpack-mastodon-modal.php:
```diff
@@ -0,0 +1,195 @@
+	 */
+	public static function should_display_modal() {
+		if (
+			// phpcs:disable WordPress.Security.NonceVerification.Recommended -- we only check if the query strings exist, we do not use them.
```
Not really a good reason. p9dueE-4z8-p2#we-re-not-using-the-submitted-data-just-the-fact-that-a-field-is-set

The good reason is that we only use this to display a UI, not to change anything on the site.

jeherve replied:
That's true. 3eb672f9a370f98fe8340d1b807ee7633555956c

anomiex commented on projects/plugins/jetpack/modules/sharedaddy/services/class-jetpack-mastodon-modal.php:
```diff
@@ -0,0 +1,195 @@
+		<div class="sd-mastodon-modal__logo">
+			<?php echo self::$svg; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped -- SVG is safe. ?>
+		</div>
+		<form action="" method="post">
```
The `action` attribute [shouldn't be empty in HTML5](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fs-action). If you want it to submit back to the same page, it should just be omitted.

Back in HTML 4 it was required and an empty string was ok, but the doctype used here is an HTML5 doctype.

jeherve replied:
Done in c1197e23a8ab9573874676c65a4e3718dc30c91e

anomiex commented on projects/plugins/jetpack/modules/sharedaddy/sharing-sources.php:
```diff
@@ -153,6 +153,46 @@ public function get_share_title( $post_id ) {
+		 */
+		$tag_list = (string) apply_filters( 'jetpack_sharing_tag_list', implode( ', ', $tags ), $post_id, $this->id );
+
+		return html_entity_decode( wp_kses( $tag_list, null ), ENT_QUOTES | ENT_SUBSTITUTE | ENT_HTML401 );
```
`wp_kses()` is documented as taking a `string[]|string` for its `$allowed_html` parameter, not `null`.

It looks like, assuming something doesn't blow up on getting passed null, that this will behave the same as the empty string, allowing a small default set of tags.

jeherve replied:
Good point. I had copied from `get_share_title`. I updated both methods in 79ca98343bbd6fc98957c1c8a3c55c1d33fae72f

anomiex commented on projects/plugins/jetpack/modules/sharedaddy/sharing-sources.php:
```diff
@@ -3041,3 +3081,125 @@ public function display_footer() {
+	 * @param int   $id       Sharing source ID.
+	 * @param array $settings Sharing settings.
+	 */
+	public function __construct( $id, array $settings ) { // phpcs:ignore Generic.CodeAnalysis.UselessOverridingMethod.Found
```
Why ignore the sniff instead of removing the unneeded overriding method?

jeherve replied:
Because that's what we do for all the other sources in this file, and I copied and pasted one of the existing services and started from that? :)

I've removed the method entirely in bbba2e508f58d8f8bb13ea0aeca7d65205e2a1dd

anomiex commented on projects/plugins/jetpack/modules/sharedaddy/sharing-sources.php:
```diff
@@ -3041,3 +3081,125 @@ public function display_footer() {
+		 * @module sharedaddy
+		 * @since $$next-version$$
+		 *
+		 * @param string  $share_url The default message that gets posted to Mastodon. This will get rawurlencoded.
```
Why mention "This will get rawurlencoded"? Do filter hooks need to know how it's going to get processed to be sent to Mastodon?

jeherve replied:
Removed in e2b292b2a0506946cd9fd78c1dc6421f26850886

anomiex commented on projects/plugins/jetpack/modules/sharedaddy/services/class-jetpack-mastodon-modal.php:
```diff
@@ -0,0 +1,195 @@
+				id="jetpack-mastodon-instance"
+				name="jetpack-mastodon-instance"
+				placeholder="<?php echo esc_url( 'https://mastodon.social' ); ?>"
+			/>
```
I suspect a nice TODO would be to allow the user to save the value they enter for use as the default next time. But I'd be fine with that being done in a future PR.

jeherve replied:
Thta's a good point. Definitely something for a v2 though. Where would you suggest we save this? In localStorage?

anomiex replied:
Good question. When I wrote the comment I had mistakenly thought this was like Publicize where it was the site owner using it, rather than the sharing buttons where readers are re-sharing.

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into add/mastodon-sharing-button

anomiex reviewed (commented):
Gave the code a once-over.

When I try the UI on a JN site

> Drag a Mastodon sharing button to one of the sharing areas.
> * You will want with different button styles.

When I select "Official buttons" I get tofu for the Mastodon button in the settings page. Frontend looks ok though.

anomiex commented:
> When I select "Official buttons" I get tofu for the Mastodon button in the settings page.

Looks like there's special casing here that would need another case added:
https://github.com/Automattic/jetpack/blob/e8ac2a7943a87fae7dc76796cae874de305a3714/projects/plugins/jetpack/modules/sharedaddy/admin-sharing.js#L166-L172

jeherve committed:
Untofu official button's preview

See https://github.com/Automattic/jetpack/pull/28694#issuecomment-1438670355

jeherve committed:
Remove useless overriding method

See https://github.com/Automattic/jetpack/pull/28694#discussion_r1113151403

jeherve committed:
Remove extra comment.

see https://github.com/Automattic/jetpack/pull/28694#discussion_r1113160240

anomiex committed:
Remove unneeded action attribute

See https://github.com/Automattic/jetpack/pull/28694#discussion_r1113138533

jeherve committed:
Update phpcs ignore comment

See https://github.com/Automattic/jetpack/pull/28694#discussion_r1113124415

anomiex committed:
Switch to a better default allowed HTML for wp_kses

See https://github.com/Automattic/jetpack/pull/28694#discussion_r1113150305

jeherve commented:
> When I select "Official buttons" I get tofu for the Mastodon button in the settings page.

Thanks for letting me know, I had missed that. I pushed a commit to fix it.

Also, I have to ask; does "tofu" mean something specific? It's the first time I hear this used and not referring to the soy-based food. :)

anomiex commented:
> Also, I have to ask; does "tofu" mean something specific? It's the first time I hear this used and not referring to the soy-based food. :)

Sorry. In this context it refers to the little box (sometimes with numbers in it) that's used as a placeholder when no font provides the character. See https://fonts.google.com/knowledge/glossary/tofu for example.

anomiex reviewed (approved):
Ok with this as it is, so I'll hit Approve. The remaining comments IMO are ok to ignore or to leave for a followup if you want.

jeherve commented:
> Sorry. In this context it refers to the little box (sometimes with numbers in it) that's used as a placeholder when no font provides the character. See [fonts.google.com/knowledge/glossary/tofu](https://fonts.google.com/knowledge/glossary/tofu) for example.

Oh! TIL, I had no idea it was named like that. It all makes sense. Thank you!
